### DISCUSSION TITLE
Possible idea for less destructive transclusion

### DISCUSSION THREAD

**Discussion by tefkah** [nobiot/org-transclusion#65] [Ideas]
Hi nobiot, really love the package, this is exactly how I would want org-roam to function and prevents me from rewriting a ton of notes in different places.

However, the "destructive" nature of org-transclusion has been bothering me as well, and while it hasn't erased my work yet, it has come somewhat close. The most egregious thing that has happened has only been the deletion and non-reinsertion of some links, which is mildly annoying as I sometimes forgot that it did that. The problem I run into the most though, is org-transclusion simply not inserting the links, which is probably my fault since I use a lot of janky minor modes. 

So I wanted to run an idea by you to see if it's feasible/you haven't tried it yourself. AFAIK, you decided to use the destructive version (partially) because otherwise exports would not work, which I get. However I think that could be fixed:

- Continue using the transclusion keyword
- Create a new export function, `org-transclude-export`, which creates a new temporary buffer and either
-- does the "destructive" transclusion there, or, more elegantly perhaps
-- changes the `#+transclusion` to `#+include` with some modifications of `:hlevel` to `:minlevel` (this might not work that well)
- Implement the transclusion to just insert the text under the `#+tranclusion-header` when `org-transclusion-mode` is active, maybe like John Kitchin's implementation, but instead of making it `:keymap` dependent activating it whenever the minor mode is active.

I do need to give the disclaimer that I haven't looked deeply your code myself, so this might me nonsense. In any case, hope to hear from you, if you think this is worth trying I would love to help work on an implementation.

**Comment by nobiot** [comment]
Thank you for your comment and feedback. I am slowly coming back to this package (I have had to take care of some other commitments and deadlines in my other parts of life).

Your comment has come at the right time. I am prototyping some changes. Your comment is helpful to me because it confirms my concern that deleting the link might not be the best approach.

There are a couple things I will need to unpack here. Let me take a few days (hopefully only a few days) and come back to this thread. What I have done so far is still "destructive" but I might be able to avoid this and come up with an alternative---I need to see how I can incorporate your ideas.

Thank you.

**Reply by nobiot** [reply]
@ThomasJorna 
Just very quickly, here is what I have been prototyping (branch only in my local now).

Instead of using the link, I am using just `#+transclude` keyword. I was deleting the keyword ("destructive") but I guess that was for no reason. Perhaps it might give users more assurance that the link won't be deleted even when transclusion failes for any reason.

As it is a keyword, export does not include it.

See the GIF animation below. I just quickly changed the code so that the original transclude keyword will remain, and the content gets added just below it.

![2021-04-21-new-keyword-transclusion-0 0 7](https://user-images.githubusercontent.com/12507865/115601935-a9ca9f00-a2de-11eb-8293-76d2092cbb65.gif)

This is based on the `#+include` syntax. I am thinking of extending it to support source code transclusion discussed in [this issue](https://github.com/nobiot/org-transclusion/issues/57). What do you think?
cc: @stardiviner

**Reply by stardiviner** [reply]
This is looks good enough now. Supporting source code transclusion is what I want for now. Otherwise I have to add more headlines for reference and more detailed headlines. Is it possible to transclude the `#+NAME` content? So that all partial content which has `#+NAME` can be transcluded, like table etc. I don't know Org Mode whether has similar API can locate and limite the `#+NAME` partial content. This might be more extensive.

**Reply by nobiot** [reply]
@stardiviner Thanks!
Will look at the source code transclusion at some point. At the moment, I am looking at live update and reinforcing undo/redo operations. 

Re the `#+name` content, yes. See [this part of README](https://github.com/nobiot/org-transclusion#org--mode-links) and the image in Figure 3 in the [introduction](https://github.com/nobiot/org-transclusion#introduction) - it show cases transclusion of a paragraph, block quote, table, and list.

**Reply by stardiviner** [reply]
Aha, indeed, sorry, my fault, I missed this in my head.

**Reply by nobiot** [reply]
No worries :)

**Comment by nobiot** [comment]
>     * Implement the transclusion to just insert the text under the `#+tranclusion-header`

@ThomasJorna 
I think I [tried out](https://github.com/nobiot/org-transclusion/discussions/65#discussioncomment-641477) your suggestion (please correct me if I misunderstood your suggestion) 

The issue is really not export. It's the structure of the resultant Org file. To illustrate, see the GIF below.

![2021-04-23-new-keyword-transclusion-0 0 7-issue](https://user-images.githubusercontent.com/12507865/115833645-94f02780-a414-11eb-8446-d7d1aefd0dac.gif)

See the =#+transclude= keyword. Once transclusion is done, it should be part of the transcluded headline "Heading". But now it remains to be part of "First Heading" above. If you fold the first headline, the keyword disappears... 

This was my primary reason for "removing" the link (now keyword). There may be a way to delay the deletion until the transclusion actually happens.

**Reply by tefkah** [reply]
Aha interesting. First off all, thanks for your in-depth replies, I really appreciate it!

The "non-indenting" issue is indeed somewhat troublesome, but I actually thought this was the "intented" behavior of org-transclusion, as I have this issue at the moment as well. My solution is just to add `:hlevel` keyywords everywhere to get the indentation I desire. Do you have an idea of what exactly prevents heading from being indented correctly?

But this approach was indeed what I had in mind! The only thing I would change is to keep the link requirement as that is much easier to add than manually typing the link to the file and would keep the files linked in the org-roam-db. I do see how that might lead to some problems, but that is also the primary reason for why I was thinking of the custom export function, which then would delete the links in a new export-buffer so you would not have to see the links.

Do you think keeping the links is worth the hassle? They could always be inserted manually and removed before export.

**Reply by nobiot** [reply]
It's not the indentation that is the problem. It's in which headline the line/keyword will belong.
So yes, indentation is something I cannot easily fix (Org-transclusion does not play well with `org-indent-mode`). But the fact that the link/keyword can be separated from the transcluded text is the issue I am concerned with.

Switching to `#+translude` keyword has benefits of:
1. It's guaranteed to be one line
2. Keyword syntax takes care of it when indented with tabs (I think)
3. Code can become simpler (instead of dealing with a keyword and then a link = two lines)

I agree with you on the disadvantages:
1. It's harder to type
2. Org-roam does not recognize links in keywords as links

These can be fixed. I will play with this idea for now. I am also thinking of enhancing some UX for copy/paste, and inserting "transclusion"....

**Reply by nobiot** [reply]
Just to illustrate the issue of "separation" I am talking about...

You could move the entire headline ("First Heading" in the example above) to totally different location together with the translude keyword. Now the keyword is separated from the transcluded text. 

When you save the buffer, where will transclusion place the text again (save buffer remove all transcluded text and add them back again to keep the file clean of trascluded text)? It will be the new place where the First Heading headline is, not the original place. 

I don't like this type of surprise.

**Reply by tefkah** [reply]
Ah now I see what you mean! That is indeed a bit annoying, I have not tried this myself but I assume it works the way you expect it to in the current implementation, i.e. the transclusion stays were you put it? I don't really see an easy way of getting around that at the moment yeah, although for me this wouldn't really be a problem as I think of the transcluded content in relation with the rest of the headings anyway, nor do I move around my headings that often. But if you do prefer to manage bigger files with transcluded content this might be a problem.

Thanks for your efforts, any chance you could push your changes to another branch of this repo so I could have a look? I would like to get familiar with the structure of the package so I can help out in a more concrete manner, but if the codebase is changing considerably it might not be very productive to look at it now :sweat_smile:

**Reply by nobiot** [reply]
Yes. Will push a working branch soon-ish then. Thank you!

**Reply by nobiot** [reply]
@ThomasJorna 
As requested, a new dev branch ([dev/0.1.0](https://github.com/nobiot/org-transclusion/pull/69)) is pushed. Not stable, not documented well, but workable.

See also PR #69 for a quick summary of major breaking changes in it.

**Reply by tefkah** [reply]
Fantastic! I will have a look at it during this week.
